File: frontend/src/lib/api/captable.ts
import { apiFetch } from "@/lib/api";
import type {
  ShareClass,
  Stakeholder,
  EquityEvent,
  CapTableSnapshot,
  CapTableEvolutionEntry,
  CapTableKPIs,
} from "@/types";

// ── Share Classes ────────────────────────────────────────────

export async function fetchShareClasses(
  companyId: string
): Promise<ShareClass[]> {
  return apiFetch<ShareClass[]>(`/captable/${companyId}/share-classes`);
}

export async function createShareClass(
  companyId: string,
  data: { name: string; votes_per_share?: number; liquidation_preference?: string; seniority?: number }
): Promise<ShareClass> {
  return apiFetch<ShareClass>(`/captable/${companyId}/share-classes`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function deleteShareClass(id: string): Promise<void> {
  await apiFetch<void>(`/captable/share-classes/${id}`, { method: "DELETE" });
}

// ── Stakeholders ─────────────────────────────────────────────

export async function fetchStakeholders(
  companyId: string
): Promise<Stakeholder[]> {
  return apiFetch<Stakeholder[]>(`/captable/${companyId}/stakeholders`);
}

export async function createStakeholder(
  companyId: string,
  data: {
    name: string;
    type: string;
    email?: string;
    phone?: string;
    entity_name?: string;
    contact_person?: string;
    partner_emails?: string;
    linkedin_url?: string;
    notes?: string;
  }
): Promise<Stakeholder> {
  return apiFetch<Stakeholder>(`/captable/${companyId}/stakeholders`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateStakeholder(
  id: string,
  data: Partial<{
    name: string;
    type: string;
    email: string;
    phone: string;
    entity_name: string;
    contact_person: string;
    partner_emails: string;
    linkedin_url: string;
    notes: string;
  }>
): Promise<Stakeholder> {
  return apiFetch<Stakeholder>(`/captable/stakeholders/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export async function deleteStakeholder(id: string): Promise<void> {
  await apiFetch<void>(`/captable/stakeholders/${id}`, { method: "DELETE" });
}

// ── Equity Events ────────────────────────────────────────────

export async function fetchEquityEvents(
  companyId: string
): Promise<EquityEvent[]> {
  return apiFetch<EquityEvent[]>(`/captable/${companyId}/events`);
}

export async function createEquityEvent(
  companyId: string,
  data: {
    name: string;
    event_type: string;
    date?: string;
    pre_money_valuation?: number;
    amount_raised?: number;
    price_per_share?: number;
    total_shares_after?: number;
    notes?: string;
    allocations: {
      stakeholder_id: string;
      share_class_id: string;
      shares: number;
      amount_invested?: number;
      ownership_pct?: number;
    }[];
  }
): Promise<EquityEvent> {
  return apiFetch<EquityEvent>(`/captable/${companyId}/events`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function deleteEquityEvent(id: string): Promise<void> {
  await apiFetch<void>(`/captable/events/${id}`, { method: "DELETE" });
}

// ── Cap Table Views ──────────────────────────────────────────

export async function fetchCapTableKPIs(
  companyId: string
): Promise<CapTableKPIs> {
  return apiFetch<CapTableKPIs>(`/captable/${companyId}/kpis`);
}

export async function fetchCapTableSnapshot(
  companyId: string
): Promise<CapTableSnapshot> {
  return apiFetch<CapTableSnapshot>(`/captable/${companyId}/snapshot`);
}

export async function fetchCapTableEvolution(
  companyId: string
): Promise<CapTableEvolutionEntry[]> {
  return apiFetch<CapTableEvolutionEntry[]>(
    `/captable/${companyId}/evolution`
  );
}

File: backend/src/app/schemas/captable.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


# ── Share Classes ────────────────────────────────────────────

class ShareClassCreate(BaseModel):
    name: str
    votes_per_share: int = 1
    liquidation_preference: str | None = None
    seniority: int = 0


class ShareClassRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    votes_per_share: int
    liquidation_preference: str | None
    seniority: int
    created_at: datetime | None = None


# ── Stakeholders ─────────────────────────────────────────────

class StakeholderCreate(BaseModel):
    name: str
    email: str | None = None
    phone: str | None = None
    type: str  # founder | employee | angel | vc | other
    entity_name: str | None = None
    contact_person: str | None = None
    partner_emails: str | None = None
    linkedin_url: str | None = None
    notes: str | None = None


class StakeholderUpdate(BaseModel):
    name: str | None = None
    email: str | None = None
    phone: str | None = None
    type: str | None = None
    entity_name: str | None = None
    contact_person: str | None = None
    partner_emails: str | None = None
    linkedin_url: str | None = None
    notes: str | None = None


class StakeholderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    email: str | None
    phone: str | None
    type: str
    entity_name: str | None
    contact_person: str | None
    partner_emails: str | None
    linkedin_url: str | None
    notes: str | None
    created_at: datetime | None = None


# ── Allocations ──────────────────────────────────────────────

class AllocationCreate(BaseModel):
    stakeholder_id: str
    share_class_id: str
    shares: int
    amount_invested: float | None = None
    ownership_pct: float = 0.0
    notes: str | None = None


class AllocationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    stakeholder_id: str
    stakeholder_name: str | None = None
    share_class_id: str
    share_class_name: str | None = None
    shares: int
    amount_invested: float | None
    ownership_pct: float
    notes: str | None


# ── Equity Events ────────────────────────────────────────────

class EquityEventCreate(BaseModel):
    name: str
    event_type: str  # incorporation | funding_round | grant | secondary | conversion
    date: str | None = None  # ISO date string
    pre_money_valuation: float | None = None
    amount_raised: float | None = None
    price_per_share: float | None = None
    total_shares_after: int | None = None
    notes: str | None = None
    allocations: list[AllocationCreate] = []


class EquityEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    event_type: str
    date: datetime | None
    pre_money_valuation: float | None
    amount_raised: float | None
    price_per_share: float | None
    total_shares_after: int | None
    notes: str | None
    allocations: list[AllocationRead] = []
    created_at: datetime | None = None


# ── Cap Table Snapshot ───────────────────────────────────────

class CapTableRow(BaseModel):
    stakeholder: StakeholderRead
    shares_by_class: dict[str, int]
    total_shares: int
    ownership_pct: float
    total_invested: float


class CapTableSnapshot(BaseModel):
    rows: list[CapTableRow]
    total_shares: int
    share_classes: list[ShareClassRead]


class CapTableEvolutionEntry(BaseModel):
    event: EquityEventRead
    snapshot: list[CapTableRow]


# ── KPI Summary ──────────────────────────────────────────────

class CapTableKPIs(BaseModel):
    last_valuation: float | None
    post_money_valuation: float | None
    total_raised: float
    total_shareholders: int
    founder_ownership_pct: float
    total_shares: int
    rounds_count: int
    last_round_name: str | None
